File: egottol/engines/analog/noise.py
"""Analog noise models for circuit simulation traces."""

from __future__ import annotations

from typing import Optional, Union

import numpy as np

KB = 1.380649e-23
Q_E = 1.602176634e-19


def thermal_noise(
    resistance: float,
    temperature: float = 300.0,
    bandwidth: float = 1.0,
    n_samples: int = 1,
    rng: Optional[np.random.Generator] = None,
) -> np.ndarray:
    """
    Johnson-Nyquist thermal noise voltage RMS: V_n = sqrt(4 k T R BW).

    Returns zero-mean Gaussian samples with that standard deviation.
    """
    r = max(float(resistance), 1e-18)
    bw = max(float(bandwidth), 1e-18)
    v_rms = np.sqrt(4.0 * KB * temperature * r * bw)
    gen = rng or np.random.default_rng()
    return gen.normal(0.0, v_rms, n_samples)


def flicker_noise(
    n_samples: int,
    dt: float,
    corner_freq: float = 1.0,
    amplitude: float = 1e-6,
    rng: Optional[np.random.Generator] = None,
) -> np.ndarray:
    """
    1/f (flicker) noise via spectral shaping of white noise.

    PSD ~ 1/f below corner_freq; generates a time-domain trace of length n_samples.
    """
    gen = rng or np.random.default_rng()
    n = max(int(n_samples), 1)
    dt = max(float(dt), 1e-18)
    freqs = np.fft.rfftfreq(n, dt)
    white = gen.normal(0.0, 1.0, freqs.size) + 1j * gen.normal(0.0, 1.0, freqs.size)
    psd = np.ones_like(freqs, dtype=float)
    mask = freqs > 0
    psd[mask] = 1.0 / np.sqrt(freqs[mask] / max(corner_freq, 1e-18))
    psd[0] = 0.0
    shaped = white * psd
    trace = np.fft.irfft(shaped, n=n)
    trace = trace / max(np.std(trace), 1e-18) * amplitude
    return trace


def add_noise_to_trace(
    trace: np.ndarray,
    thermal_r: float = 0.0,
    temperature: float = 300.0,
    bandwidth: float = 1.0,
    flicker_amp: float = 0.0,
    dt: float = 1e-3,
    flicker_corner: float = 1.0,
    rng: Optional[np.random.Generator] = None,
) -> np.ndarray:
    """Add thermal and optional flicker noise to a voltage/current trace."""
    y = np.asarray(trace, dtype=float).copy()
    n = y.size
    gen = rng or np.random.default_rng()

    if thermal_r > 0.0:
        y += thermal_noise(thermal_r, temperature, bandwidth, n_samples=n, rng=gen)

    if flicker_amp > 0.0:
        y += flicker_noise(n, dt, corner_freq=flicker_corner, amplitude=flicker_amp, rng=gen)

    return y

File: core/analog/noise.cpp
#include "noise.h"
#include <cmath>
#include <complex>
#include <algorithm>

namespace deepiri {

std::vector<double> thermalNoise(double resistance, double temperature, double bandwidth,
                                  size_t nSamples, std::mt19937_64& rng) {
    double r = std::max(resistance, 1e-18);
    double bw = std::max(bandwidth, 1e-18);
    double vRms = std::sqrt(4.0 * kBoltzmann * temperature * r * bw);
    std::normal_distribution<double> dist(0.0, vRms);
    std::vector<double> out(std::max<size_t>(nSamples, 1));
    for (auto& v : out) {
        v = dist(rng);
    }
    return out;
}

std::vector<double> thermalNoise(double resistance, double temperature, double bandwidth,
                                  size_t nSamples) {
    std::random_device rd;
    std::mt19937_64 rng(rd());
    return thermalNoise(resistance, temperature, bandwidth, nSamples, rng);
}

std::vector<double> flickerNoise(size_t nSamples, double dt, double cornerFreq,
                                  double amplitude, std::mt19937_64& rng) {
    size_t n = std::max<size_t>(nSamples, 1);
    double dtSafe = std::max(dt, 1e-18);
    double cornerSafe = std::max(cornerFreq, 1e-18);

    size_t m = n / 2 + 1;
    std::vector<double> freqs(m);
    for (size_t k = 0; k < m; ++k) {
        freqs[k] = static_cast<double>(k) / (static_cast<double>(n) * dtSafe);
    }

    std::normal_distribution<double> dist(0.0, 1.0);
    std::vector<std::complex<double>> shaped(m);
    for (size_t k = 0; k < m; ++k) {
        std::complex<double> white(dist(rng), dist(rng));
        double psd = 1.0;
        if (k > 0) {
            psd = 1.0 / std::sqrt(freqs[k] / cornerSafe);
        } else {
            psd = 0.0;
        }
        shaped[k] = white * psd;
    }

    std::vector<std::complex<double>> full(n);
    for (size_t k = 0; k < n; ++k) {
        if (k < m) {
            full[k] = shaped[k];
        } else {
            full[k] = std::conj(shaped[n - k]);
        }
    }

    std::vector<double> trace(n, 0.0);
    const double twoPiOverN = 2.0 * M_PI / static_cast<double>(n);
    for (size_t j = 0; j < n; ++j) {
        std::complex<double> acc(0.0, 0.0);
        for (size_t k = 0; k < n; ++k) {
            double angle = twoPiOverN * static_cast<double>(k) * static_cast<double>(j);
            acc += full[k] * std::complex<double>(std::cos(angle), std::sin(angle));
        }
        trace[j] = acc.real() / static_cast<double>(n);
    }

    double mean = 0.0;
    for (double v : trace) mean += v;
    mean /= static_cast<double>(n);
    double variance = 0.0;
    for (double v : trace) variance += (v - mean) * (v - mean);
    variance /= static_cast<double>(n);
    double stddev = std::max(std::sqrt(variance), 1e-18);

    for (double& v : trace) {
        v = v / stddev * amplitude;
    }
    return trace;
}

std::vector<double> flickerNoise(size_t nSamples, double dt, double cornerFreq, double amplitude) {
    std::random_device rd;
    std::mt19937_64 rng(rd());
    return flickerNoise(nSamples, dt, cornerFreq, amplitude, rng);
}

std::vector<double> addNoiseToTrace(const std::vector<double>& trace, double thermalR,
                                     double temperature, double bandwidth, double flickerAmp,
                                     double dt, double flickerCorner, std::mt19937_64& rng) {
    std::vector<double> y = trace;
    size_t n = y.size();

    if (thermalR > 0.0) {
        auto tn = thermalNoise(thermalR, temperature, bandwidth, n, rng);
        for (size_t i = 0; i < n; ++i) {
            y[i] += tn[i];
        }
    }

    if (flickerAmp > 0.0) {
        auto fn = flickerNoise(n, dt, flickerCorner, flickerAmp, rng);
        for (size_t i = 0; i < n; ++i) {
            y[i] += fn[i];
        }
    }

    return y;
}

std::vector<double> addNoiseToTrace(const std::vector<double>& trace, double thermalR,
                                     double temperature, double bandwidth, double flickerAmp,
                                     double dt, double flickerCorner) {
    std::random_device rd;
    std::mt19937_64 rng(rd());
    return addNoiseToTrace(trace, thermalR, temperature, bandwidth, flickerAmp, dt, flickerCorner, rng);
}

}
